benches/bin/rpc.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::thread;
+use std::str::FromStr;
+use std::time::{Instant, Duration};
+use std::net::{SocketAddr, IpAddr};
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc::{Sender, Receiver, channel};
+
+use kvproto::tikvpb_grpc::*;
+use kvproto::coprocessor::{Request, Response};
+use kvproto::raft_serverpb::{RaftMessage, Done, SnapshotChunk};
+use kvproto::kvrpcpb::*;
+
+use futures::{Future, Stream, future};
+use grpc::{Server as GrpcServer, ServerBuilder, Environment, ChannelBuilder, RpcContext,
+           UnarySink, RequestStream, ClientStreamingSink};
+
+use tikv::server::{RaftClient, Config};
+
+use super::print_result;
+
+// Copy from std::time;
+const NANOS_PER_SEC: u32 = 1_000_000_000;
+
+struct Inner {
+    name: String,
+    counter: Arc<AtomicUsize>,
```

Comment:
Mutex is for mut accessing `Inner`. This counter is shared between threads and I want it outlive `MutexGuard` so that it does not need to hold the lock for the whole RPC call.